fix(button): ensure link styles don't override a link styled as button

owensgit opened the pull request:
# Description

~Th `lg-link` mixin used `!important` to apply colours to links inside
one of the `lg-variant` classes (such as info, success, error, etc).
However, this caused issues with the `lg-button` directive when used to
style a link as a button, because the global link styles use `lg-link`.
This is fixed by making `!important` an opt in paramter of the `lg-link`
mixin, so that the variant styles can still use it and apply `!important`,
but that won't affect any other uses of the `lg-link` mixin.~

**Edit:**

Following the discussion on the use of `!important` for the links inside `lg-variant`, we have decided to remove `!important` completely from the `lg-link` mixin.


This change also edits the button story, so that it displays both a normal button and a link styled as a button (separate commit).



Storybook link: https://deploy-preview-336--legal-and-general-canopy.netlify.app/?path=/story/components-button--primary-button

**Before:**
<img width="252" alt="Screenshot 2021-04-27 at 14 57 37" src="https://user-images.githubusercontent.com/1943532/116257202-bed68080-a76b-11eb-9e8c-14820a8bb8e4.png">



**After:**
<img width="247" alt="Screenshot 2021-04-27 at 14 57 04" src="https://user-images.githubusercontent.com/1943532/116256049-bf224c00-a76a-11eb-8082-b08314df766b.png">



# Checklist:

- [x] The commit messages follow the [convention for this project](./blob/master/docs/CONTRIBUTING.md#conventional-commits)
- [ ] I have provided an adequate amount of test coverage
- [ ] I have added the functionality to the [test app](./blob/master/docs/CONTRIBUTING.md#build-test-application)
- [ ] I have provided a story in storybook to document the changes
- [ ] I have provided documentation in the notes section of the story
- [ ] I have added any new public feature modules to public-api.ts

elenagarrone commented:
I'm wondering if maybe it should be the button to have the colours with `important` if it's a link?

In the current implementation I would probably try to find a better name for the variable? or maybe add a comment to explain what's the usage of that argument for people wanting to use the mixin?

owensgit commented:
@elenagarrone 

> I'm wondering if maybe it should be the button to have the colours with `important` if it's a link?

`!important` was originally added just because of  the `lg-variant` directive - I think just to follow what the margin/padding directives do - they use `!important` on their styles to ensure that they always work. At some point in the `next` branch, `!important` got added to the global link styles, which I don't think is correct anyway - and is what caused this issue with the button. So this approach essentially reverts back to the original setup whereby any links inside `lg-variant` only had `!important` applied

I suppose one question is, do we want `!important` on the `lg-variant` directive styles? If so, I think we need to keep this approach, because it only applies `!important` for link colours inside `lg-variant`.

If we don't want to use `!important` for `lg-variant`, then we could just remove it completely - it actually seems to work fine without it.

Either way, I don't think we need the button colours to use `!important` - again because the original use of `!important` was only for links inside of `lg-variant`.

My inclination is to remove `!important` entirely.

Hope that makes sense, happy to jump on a call to discuss 🙂

elenagarrone commented:
> @elenagarrone
> 
> > I'm wondering if maybe it should be the button to have the colours with `important` if it's a link?
> 
> `!important` was originally added just because of the `lg-variant` directive - I think just to follow what the margin/padding directives do - they use `!important` on their styles to ensure that they always work. At some point in the `next` branch, `!important` got added to the global link styles, which I don't think is correct anyway - and is what caused this issue with the button. So this approach essentially reverts back to the original setup whereby any links inside `lg-variant` had `!important` applied and nowhere else.
> 
> I suppose one question is, do we want `!important` on the `lg-variant` directive styles? If so, I think we need to keep this approach, or similar, because it only applies `!important` for link colours inside `lg-variant`.
> 
> If we don't want to use `!important` for `lg-variant`, then we could just remove it completely - it actually seems to work fine without it.
> 
> Either way, I don't think we need the button colours to use `!important` - again because the original use of `!important` was only for links inside of `lg-variant`.
> 
> My inclination is to remove `!important` entirely.
> 
> Hope that makes sense, happy to jump on a call to discuss 🙂

Yes, thanks for the explanation, I agree 👍 
I thought `!important` was added to the variants because it was breaking something?

owensgit commented:
> From memory we only started to add !important to deal with apps migrating to Canopy, to make sure that the Canopy styles won over #49 (@andy-polhill)

Yes that's right, we added `!important` to `lgMargin` and `lgPadding` because of that reason.  I then added it to `lgVariant` just by way of following a pattern. I think it was removed with the introduction of the `lg-link` mixin, but then added back in a subsequent fix. As mentioned below, I don't think we actually need it though.

> I thought !important was added to the variants because it was breaking something? (@elenagarrone)

I thought that too - which I why I created that mixin argument - but then I removed `!important` all together, tested it Storybook and in-situ in my app, it seems to work fine. @elenagarrone Do you remember if there were any issues at your end at all?

> However it does feel like we are falling into the trap of a specificity war. (@andy-polhill)

Yes I agree, I think given the above testing without `!important` shows that we can probably remove it in this instance, and then possibly look at removing them from `lgMargin` and `lgPadding` if they Angular now takes care of that for us.

elenagarrone commented:
> > I thought !important was added to the variants because it was breaking something? (@elenagarrone)
> 
> I thought that too - which I why I created the mixin argument in this PR - but then I removed `!important` all together, tested it Storybook and in-situ in my app, it seems to work fine. @elenagarrone Do you remember if there were any issues at your consuming app at all?

I thought it was because we had to override the generic classes setting colours on the components but now that I think about it I don't think it would matter, right? 

Did you not open this to fix any bug? or did you just noticed that the `!important` was missing?

owensgit commented:
@elenagarrone @andy-polhill @paul-request  - Thanks for the great discussion on this, as agreed, I've removed `!important` from the link styles entirely, even for `lg-variant`.

andy-polhill reviewed (approved):
Good fix @owensgit thanks for chasing this one down.
